### PR TITLE
Fix S4158: "Empty collections" raises false positive when analyzed me…

### LINKED ISSUE #767: Fix S4158: "Empty collections" raises false positive when analyzed method has too many branches
A false positive S4158 is thrown on the line: 
```csharp 
datagridview[columnname, datagridviewrow.Index].Style.BackColor = columnnamecolordictionary[columnname];
```

```csharp
public static void ColorRow(UserControl usercontrol, DataGridViewRow datagridviewrow, List<string> columnnamelist)
{
    string name = usercontrol.Name.Replace("UserControl", string.Empty);
    DataGridView datagridview = datagridviewrow.DataGridView;

    Dictionary<string, Color> columnnamecolordictionary = new Dictionary<string, Color>();

    for (int i = 0; i < datagridviewrow.DataGridView.Columns.Count; i++)
    {
        foreach (string columnname in columnnamelist)
        {
            if ((columnname == name || columnname == "*") && columnname == datagridviewrow.DataGridView.Columns[i].Name && datagridviewrow.Cells[i].Value != DBNull.Value && !string.IsNullOrWhiteSpace(datagridviewrow.Cells[i].Value.ToString()))
            {
                if (columnname == "=")
                {
                    // do nothing
                }

                columnnamecolordictionary.Add(columnname, Color.AliceBlue);
            }
        }
    }

    foreach (string columnname in columnnamecolordictionary.Keys)
    {
        if (datagridview.Columns.Contains(columnname))
        {
            datagridview[columnname, datagridviewrow.Index].Style.BackColor = columnnamecolordictionary[columnname];
        }
    }
}
```

### PR DESCRIPTION
**Pull request by valhristov**
…thod has too many branches

Fix #767 

See ticket comments for more info

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EmptyCollectionsShouldNotBeEnumerated.cs [hunk 1]
```diff
@@ -101,9 +101,16 @@ private void CheckForEmptyCollectionAccess(CSharpExplodedGraph explodedGraph, Sy
...
+            EventHandler explorationEnded =
+                (sender, args) => emptyCollections.Except(nonEmptyCollections)
+                    .Select(node => Diagnostic.Create(rule, node.GetLocation()))
+                    .ToList()
+                    .ForEach(d => context.ReportDiagnosticWhenActive(d));
+
             EventHandler<CollectionAccessedEventArgs> collectionAccessedHandler = (sender, args) =>
                 (args.IsEmpty ? emptyCollections : nonEmptyCollections).Add(args.Node);
 
+            explodedGraph.ExplorationEnded += explorationEnded;
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EmptyCollectionsShouldNotBeEnumerated.cs]: ExplorationEnded is not raised if the execution limit is reached. We will not report in complex methods, exactly as the "Condition evaluates to constant" rule does.

Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EmptyCollectionsShouldNotBeEnumerated.cs]: That's not part of this PR/ticket but could you check the other rules to see if this is present where needed?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/EmptyCollectionsShouldNotBeEnumerated.cs [hunk 2]
```diff
@@ -723,4 +723,52 @@ public Program(string id)
             if (string.IsNullOrEmpty(id)) throw new ArgumentNullException(nameof(id));
         }
     }
+
+    // Large CFG that causes the exploded graph to hit the exploration limit
+    // See https://github.com/SonarSource/sonar-csharp/issues/767 (comments!)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/EmptyCollectionsShouldNotBeEnumerated.cs]: Probably better to link directly the comment (https://github.com/SonarSource/sonar-csharp/issues/767#issuecomment-362338301)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EmptyCollectionsShouldNotBeEnumerated.cs (hunk 1 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EmptyCollectionsShouldNotBeEnumerated.cs]: I already checked. I think they are fine.

**Pull request merged**